File: src/lib/simulation.ts
import type { InputValues } from './config';
import { DEFAULT_VARIABLE_KEYS, TRACKED_METRICS, METRIC_LABELS } from './config';
import type { FinancialResults } from './models';
import { calculate } from './models';

export interface VariableRange {
  key: string;
  label: string;
  low: number;
  base: number;
  high: number;
  enabled: boolean;
}

export function buildDefaultRanges(baseInputs: InputValues, spread = 0.20): VariableRange[] {
  return DEFAULT_VARIABLE_KEYS.map((spec) => {
    const baseVal = baseInputs[spec.key as keyof InputValues] as number;
    const low = Math.round(baseVal * (1 - spread) * 10000) / 10000;
    const high = Math.round(baseVal * (1 + spread) * 10000) / 10000;
    return {
      key: spec.key,
      label: spec.label,
      low,
      base: Math.round(baseVal * 10000) / 10000,
      high,
      enabled: true,
    };
  });
}

function triangular(low: number, mode: number, high: number): number {
  const u = Math.random();
  const fc = (mode - low) / (high - low);
  return u < fc
    ? low + Math.sqrt(u * (high - low) * (mode - low))
    : high - Math.sqrt((1 - u) * (high - low) * (high - mode));
}

function mean(arr: number[]): number {
  let sum = 0;
  for (let i = 0; i < arr.length; i++) sum += arr[i];
  return sum / arr.length;
}

function std(arr: number[], avg: number): number {
  let sum = 0;
  for (let i = 0; i < arr.length; i++) {
    const d = arr[i] - avg;
    sum += d * d;
  }
  return Math.sqrt(sum / arr.length);
}

function percentile(sorted: number[], p: number): number {
  const idx = (p / 100) * (sorted.length - 1);
  const lo = Math.floor(idx);
  const hi = Math.ceil(idx);
  if (lo === hi) return sorted[lo];
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (idx - lo);
}

export interface SummaryRow {
  metric: string;
  label: string;
  mean: number;
  std: number;
  p5: number;
  p25: number;
  median: number;
  p75: number;
  p95: number;
  probLoss: number;
}

export interface SimulationResults {
  nIterations: number;
  metricArrays: Record<string, number[]>;
  summaryRows: SummaryRow[];
}

export function runSimulation(
  baseInputs: InputValues,
  variableRanges: VariableRange[],
  nIterations = 5000,
): SimulationResults {
  const arrays: Record<string, number[]> = {};
  for (const m of TRACKED_METRICS) {
    arrays[m] = new Array(nIterations);
  }

  const activeRanges = variableRanges.filter((vr) => vr.enabled);

  for (let i = 0; i < nIterations; i++) {
    const sampled = { ...baseInputs };

    for (const vr of activeRanges) {
      const lo = Math.min(vr.low, vr.base);
      const hi = Math.max(vr.high, vr.base);
      const mode = Math.max(lo, Math.min(vr.base, hi));
      const value = lo === hi ? lo : triangular(lo, mode, hi);
      (sampled as Record<string, number>)[vr.key] = value;
    }

    const result: FinancialResults = calculate(sampled);
    for (const metric of TRACKED_METRICS) {
      arrays[metric][i] = result[metric as keyof FinancialResults] as number;
    }
  }

  const summaryRows: SummaryRow[] = TRACKED_METRICS.map((metric) => {
    const arr = arrays[metric];
    const avg = mean(arr);
    const sorted = [...arr].sort((a, b) => a - b);
    const negCount = arr.reduce((c, v) => c + (v < 0 ? 1 : 0), 0);
    return {
      metric,
      label: METRIC_LABELS[metric] ?? metric,
      mean: avg,
      std: std(arr, avg),
      p5: percentile(sorted, 5),
      p25: percentile(sorted, 25),
      median: percentile(sorted, 50),
      p75: percentile(sorted, 75),
      p95: percentile(sorted, 95),
      probLoss: (negCount / arr.length) * 100,
    };
  });

  return { nIterations, metricArrays: arrays, summaryRows };
}

File: simulation.py
"""Motor de simulación Monte Carlo para escenarios de producción de café.

Ejecuta N iteraciones del modelo financiero, muestreando cada entrada
incierta de una distribución triangular. Devuelve SimulationResults
con arreglos por métrica y estadísticas resumidas.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import numpy as np
import pandas as pd

from models import FinancialResults, calculate


@dataclass
class VariableRange:
    key: str
    label: str
    low: float
    base: float
    high: float
    enabled: bool = True


DEFAULT_VARIABLE_KEYS: list[dict[str, str]] = [
    {"key": "cherry_yield_per_ha", "label": "Rendimiento de Cereza (lbs/ha)"},
    {"key": "ratio_honey", "label": "Ratio Cereza → Honey (%)"},
    {"key": "ratio_natural", "label": "Ratio Cereza → Natural (%)"},
    {"key": "ratio_pilado", "label": "Ratio Cereza → Pilado (%)"},
    {"key": "price_cereza", "label": "Precio Cereza (B/./lb)"},
    {"key": "price_honey", "label": "Precio Seco Honey (B/./lb)"},
    {"key": "price_natural", "label": "Precio Seco Natural (B/./lb)"},
    {"key": "price_pilado", "label": "Precio Seco Pilado (B/./lb)"},
    {"key": "fertilizer", "label": "Costo de Fertilizante (B/.)"},
    {"key": "pesticide", "label": "Control de Plagas (B/.)"},
    {"key": "seasonal_workers", "label": "Trabajadores Temporales"},
    {"key": "seasonal_daily_wage", "label": "Jornal Diario Temporal (B/.)"},
    {"key": "harvest_days", "label": "Días de Cosecha"},
    {"key": "cost_honey_lb", "label": "Costo Proceso Honey (B/./lb)"},
    {"key": "cost_natural_lb", "label": "Costo Proceso Natural (B/./lb)"},
    {"key": "cost_pilado_lb", "label": "Costo Proceso Pilado (B/./lb)"},
    {"key": "fuel", "label": "Combustible y Energía (B/.)"},
    {"key": "transport", "label": "Transporte y Logística (B/.)"},
]


def build_default_ranges(
    base_inputs: dict[str, Any],
    spread: float = 0.20,
) -> list[VariableRange]:
    ranges: list[VariableRange] = []
    for spec in DEFAULT_VARIABLE_KEYS:
        key = spec["key"]
        base_val = float(base_inputs.get(f"in_{key}", base_inputs.get(key, 0)))
        low = base_val * (1.0 - spread)
        high = base_val * (1.0 + spread)
        ranges.append(VariableRange(
            key=key, label=spec["label"],
            low=round(low, 4), base=round(base_val, 4), high=round(high, 4),
            enabled=True,
        ))
    return ranges


TRACKED_METRICS: list[str] = [
    "total_revenue",
    "total_costs",
    "net_profit",
    "margin",
    "cost_per_lb_cherry",
    "total_cherry",
    "honey_output_lbs",
    "natural_output_lbs",
    "pilado_output_lbs",
    "rev_per_ha",
    "profit_per_ha",
]

METRIC_LABELS: dict[str, str] = {
    "total_revenue": "Ingresos Totales (B/.)",
    "total_costs": "Costos Totales (B/.)",
    "net_profit": "Ganancia Neta (B/.)",
    "margin": "Margen de Ganancia (%)",
    "cost_per_lb_cherry": "Costo por lb Cereza (B/.)",
    "total_cherry": "Total Cereza (lbs)",
    "honey_output_lbs": "Producción Honey (lbs)",
    "natural_output_lbs": "Producción Natural (lbs)",
    "pilado_output_lbs": "Producción Pilado (lbs)",
    "rev_per_ha": "Ingresos por Hectárea (B/.)",
    "profit_per_ha": "Ganancia por Hectárea (B/.)",
}


@dataclass
class SimulationResults:
    n_iterations: int
    metric_arrays: dict[str, np.ndarray]
    summary_df: pd.DataFrame


def run_simulation(
    base_inputs: dict[str, Any],
    variable_ranges: list[VariableRange],
    n_iterations: int = 5000,
    seed: int | None = None,
) -> SimulationResults:
    rng = np.random.default_rng(seed)

    arrays: dict[str, np.ndarray] = {
        m: np.empty(n_iterations, dtype=np.float64) for m in TRACKED_METRICS
    }

    active_ranges = [vr for vr in variable_ranges if vr.enabled]

    for i in range(n_iterations):
        sampled = dict(base_inputs)

        for vr in active_ranges:
            low = min(vr.low, vr.base)
            high = max(vr.high, vr.base)
            mode = max(low, min(vr.base, high))
            if low == high:
                value = low
            else:
                value = rng.triangular(low, mode, high)
            sampled[f"in_{vr.key}"] = value

        result: FinancialResults = calculate(sampled)
        result_dict = result.as_dict()
        for metric in TRACKED_METRICS:
            arrays[metric][i] = result_dict[metric]

    summary_rows: list[dict[str, Any]] = []
    for metric in TRACKED_METRICS:
        arr = arrays[metric]
        summary_rows.append({
            "Métrica": METRIC_LABELS.get(metric, metric),
            "Media": np.mean(arr),
            "Desv. Est.": np.std(arr),
            "Percentil 5": np.percentile(arr, 5),
            "Percentil 25": np.percentile(arr, 25),
            "Mediana": np.percentile(arr, 50),
            "Percentil 75": np.percentile(arr, 75),
            "Percentil 95": np.percentile(arr, 95),
            "P(< 0)": np.mean(arr < 0) * 100,
        })

    return SimulationResults(
        n_iterations=n_iterations,
        metric_arrays=arrays,
        summary_df=pd.DataFrame(summary_rows),
    )
